Code/Engine/RendererCore/Pipeline/Implementation/RenderPipeline.cpp

```diff
@@ -1242,4 +1323,150 @@ void ezRenderPipeline::CreateDgmlGraph(ezDGMLGraph& graph)
   }
 }
 
+ezRasterizerView* ezRenderPipeline::PrepareOcclusionCulling(const ezFrustum& frustum, const ezView& view)
+{
+  if (!cvar_SpatialCullingOcclusionEnable)
+    return nullptr;
+
+  if (!ezSystemInformation::Get().GetCpuFeatures().IsAvx1Available())
+    return nullptr;
+
+  ezRasterizerView* pRasterizer = nullptr;
+
+  // extract all occlusion geometry from the scene
+  EZ_PROFILE_SCOPE("Occlusion::RasterizeView");
+
+  pRasterizer = g_pRasterizerViewPool->GetRasterizerView(view.GetViewport().width / 2, view.GetViewport().height / 2, (float)view.GetViewport().width / (float)view.GetViewport().height);
+  pRasterizer->SetCamera(view.GetCullingCamera());
+
+  {
+    EZ_PROFILE_SCOPE("Occlusion::FindOccluders");
+
+    ezSpatialSystem::QueryParams queryParams;
+    queryParams.m_uiCategoryBitmask = ezDefaultSpatialDataCategories::OcclusionStatic.GetBitmask() | ezDefaultSpatialDataCategories::OcclusionDynamic.GetBitmask();
+    queryParams.m_IncludeTags = view.m_IncludeTags;
+    queryParams.m_ExcludeTags = view.m_ExcludeTags;
+
+    m_VisibleObjects.Clear();
+    view.GetWorld()->GetSpatialSystem()->FindVisibleObjects(frustum, queryParams, m_VisibleObjects, {});
+  }
+
+  pRasterizer->BeginScene();
+
+  for (const ezGameObject* pObj : m_VisibleObjects)
+  {
+    ezMsgExtractOccluderData msg;
+    pObj->SendMessage(msg);
+
+    for (const auto& ed : msg.m_ExtractedOccluderData)
+    {
+      pRasterizer->AddObject(ed.m_pObject, ed.m_Transform);
+    }
+  }
+
+  pRasterizer->EndScene();
+
+  return pRasterizer;
+}
+
+void ezRenderPipeline::PreviewOcclusionBuffer(const ezRasterizerView& rasterizer, const ezView& view)
+{
+  if (!cvar_SpatialCullingOcclusionVisView || !rasterizer.HasRasterizedAnyOccluders())
```

Comment:
This function needs some serious cleanup at some point. I understand the problem and the solution is not so simple with our current infrastructure so we can leave it as is for now.

Reply:
I'm painfully aware of that.